Add Docker tutorial (#1840)
alphagov/govuk-developer-docs: source/manual/intro-to-govuk-docker.html.md

---

> title: Intro to Docker and govuk-docker

Reviewer: I think this is more an intro to Docker than an intro to `govuk-docker`

section: Docker
layout: manual_layout
parent: "/manual.html"
last_reviewed_on: 2019-07-01
review_in: 6 months
owner_slack: "#govuk-developers"
---

## Intro
If you are new to Docker here's a quick intro in the context of the GOV.UK stack. There is also a wrapper for a lot of the docker commands in govuk-docker, which we will look at later on.

### Install Docker

*   Download Docker [https://www.docker.com/get-started](https://www.docker.com/get-started)
*   Docker for Mac intro [https://docs.docker.com/docker-for-mac/](https://docs.docker.com/docker-for-mac/)

> ### Video explaining the concepts in Docker

Reviewer: I'm not sure this needs a separate heading. It may be worth renaming the above heading to `Getting Started` and add it as a bullet point

Author: I want this highlighted though so giving it more prominence is my intention.

> [https://www.youtube.com/watch?v=i7yoXqlg48M](https://www.youtube.com/watch?v=i7yoXqlg48M) - first 30 minutes

Reviewer: It would be nice if this URL had a nice readable name instead of just the Youtube URL. The title of the video or "Docker Introduction Video"

## Learning docker with content-publisher

> We'll run a few things here to get the [content-publisher] up and running in Docker. This is a convoluted example but it will help to explain the concepts involved.

Reviewer 2: Probably best not to mention the wrapper, as people don’t really know what docker is yet and the wrapper really wraps compose commands.

> We'll run a few things here to get the [content-publisher] up and running in Docker. This is a convoluted example but it will help to explain the concepts involved.

Reviewer: ```suggestion
We'll run a few things here to get [Content Publisher][content-publisher] up and running in Docker. This is a convoluted example but it will help to explain the concepts involved.

You will need to have cloned the content-publisher repo to follow along with this tutorial.
```

### Run a container from an image

> *   Make sure you have docker running, then run from the root of the [content-publisher] project:

Reviewer 2: How do people make sure Docker is running?

> *   Make sure you have docker running, then run from the root of the [content-publisher] project:

Reviewer: ```suggestion
Make sure you have docker running, then run from the root of the content-publisher project:
```

Author: I've added a second step in this section. I know they're not all quite proper points, but I like to use the bullet points to indicate a step to make in the process, even if it's just one single step per headed section. Not sure if there's a better way to convey that.

```shell

> docker run -it --rm ubuntu

Reviewer 2: Might want to explain what the flags are for, too. (Maybe a bit later.)

Reviewer: It may be worth skipping the ubuntu example and just going straight for the ruby example. I wonder if having 2 will confuse people (e.g: "why am I running ubuntu and then running ruby?")

```

#### What are Docker images and a container?

> The `ubuntu` is the "image" that we specify. Using object oriented programming as an analogy, an image is like a "class". When we "instantiate" an instance of a class, these are called "containers".

Reviewer 2: At this point, we could allude to the video, which talks about this; or we could give people an example .

Author: Seems pretty simple a concept but again adding a link to above might be worthwhile in the future.

> Images are downloaded from the Docker Registry. The containers are run as a process on our host.

Reviewer 2: The process bit is true, but we might overload people with this knowledge here.

Another image is ruby:

```shell

> docker run -it --rm ruby:2.6.3 bash

Reviewer 2: Worth clarifying the distinction between flags and arguments at this point: bash is the first arg to the ‘run’ command, and is the command to be run inside the container.

```

Here we get a ruby image and specifying which version we want. The `bash` at the end means we get a bash shell.

### Volumes

> * Add a volume by running:

Reviewer 2: This isn’t really ‘adding’ a volume - a more appropriate term would be ‘mounting’.

Reviewer: I don't think this needs to be a bullet point

```shell
docker run -it --rm -v $(pwd):/app ruby:2.6.3 bash
```

> This will map your current directory (being the root of the [content-publisher] project) to `/app` inside of the container.

Reviewer: ```suggestion
This will map your current directory (the root of the content-publisher project) to `/app` inside of the container.
```

I think we also need to explain what this means in practice. E.g: being able to access those files inside the Docker container now.

> If you were to run bundle install for the app, the next time we spin up a container all of the gems would need re-installing.

Reviewer 2: Worth explaining that it’s the familiar `cd /app; bundle install`.

> If you were to run bundle install for the app, the next time we spin up a container all of the gems would need re-installing.

Reviewer: ```suggestion
If you `cd /app` and `bundle install` you can see that the gems required for content-publisher are installed! However, anything you do here won't persist - if you were to quit the container and then re-enter in exactly the same way, all the gems would need to be re-installed again.
```

#### Persistent volumes

> Docker allows you create separate volumes for persistent data.

Reviewer: ```suggestion
We don't want to be re-installing and doing setup every time we quit and re-enter a container. Docker allows you to create separate volumes for persistent data.
```

> Docker allows you create separate volumes for persistent data.

Author: Good point. Have hopefully clarified now.

* Create a persistent volume for our gems:

```shell

> docker volume create content-publisher-bundle

Reviewer: I wonder if there's a way of indicating whether we are running commands inside or outside the docker container, in a similar way we have the current Getting Started guide which refers to $mac and $dev for VM or not VM.

```

*   Run docker with that volume:

```shell
docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle ruby:2.6.3 bash
```

> *   Then installing gems:

Reviewer: ```suggestion
*  Now we can install our gems and these will be persisted even if we exit and re-enter the container:
```

```shell
cd /app

> bundle install

Reviewer 2: I guess it might confuse, but it could be useful to illustrate what ‘persisted’ means by suggesting people exit and re-enter a container.

Author: I think people would get the idea of persistent data, but if not the person who doesn't can raise the PR. As you say this may be confusing so I'm keeping it simple for now.

```

### Specifying dependencies (node, PostgreSQL, DNS)

> * Try to run the tests:

Reviewer: ```suggestion
Now we have our gems installed, we can try and run the content-publisher tests:
```

```shell
bundle exec rake
```

 We get an error about “could not find javascript runtime”. Our Ruby image isn’t doing quite what we need, so we’re going to create our own image based off that

> *   Create a Dockerfile in the content-publisher project:

Reviewer: ```suggestion
To create our own image, we need a Dockerfile. A Dockerfile normally starts with a `FROM [image]` which bases your new image off an existing image. We can then execute other commands by prefixing them with `RUN`. Each RUN command adds a new layer. You can think of it as each step creating a new image, but we only care about the final image to come out of the last step.

Create a Dockerfile in the content-publisher project:
```

```

>     FROM ruby:2.6.3

Reviewer: I think the first Dockerfile we created only had these commands:

```
FROM ruby:2.6.3

RUN curl -sL https://deb.nodesource.com/setup_12.x | bash
RUN apt-get install nodejs && apt-get clean
```

    RUN curl -sL https://deb.nodesource.com/setup_12.x | bash

    RUN apt-get install nodejs && apt-get clean

>     RUN apt-get install -y libxss1 libappindicator1 libindicator7 && apt-get clean

Reviewer 2: This comes later, since we don't know we need Chrome yet.

    RUN wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb

    RUN dpkg -i google-chrome-stable_current_amd64.deb; apt-get -fy install

>     RUN useradd -m build

Reviewer 2: This also comes later - again with Chrome.

    USER build

    ```

> *   `docker build -t content-publisher .`

Reviewer: ```suggestion
Build our new image:

`docker build -t content-publisher .`
```

> *   So now we can run it as: `docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle content-publisher bash`

Reviewer: ```suggestion
Now we can start a container with our new image: `docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle content-publisher bash`

Each time you change the Dockerfile, you need to remember to rebuild the docker image.
```

### Postgres image/container

> *   `docker run -it --rm postgres`

Reviewer 2: Perhaps worth linking to the Postgres image on DockerHub, so it becomes clearer where they're coming from and how we know what to type here.

Author: Think I've already mentioned Docker Registry

> *   `docker ps` - to see what containers you have running

Reviewer 2: (In another tab.)

*   `docker inspect [container ID for postgres]`. Grab the IP address
*   In the content-publisher container:
    *   `export DATABASE_URL=postgres://postgres@172.17.0.3/content-publisher-dev`
    *   `export TEST_DATABASE_URL=postgres://postgres@172.17.0.3/content-publisher-test`

> *   Run the tests in content-publisher container: `bundle exec rake db:setup`

Reviewer 2: We're actually setting up the DB here.

Reviewer 2: Kind of missing some explanation about the need for the volume again here. Not sure how much we want to emphasise it, though 🤔 .

Author: Think my rewrite covers this

*   Create a volume to persist to for the data: `docker volume create content-publisher-postgres`

> *   `docker run -it --rm -v content-publisher-postgres:/var/lib/postgresql/data postgres`

Reviewer 2: We should explain how we figure out where to mount it to.

Author: I think keeping the focus of the tutorial on using Docker, as opposed to where what common system structures are is clearer at this point but again, open to changes in the future.

> *   Add `builder` user to Dockerfile (see above) then re-build content-publisher container: `docker build -t content-publisher .`

Reviewer 2: This looks like it's got mixed up with the Chrome stuff - appreciate I was going rather fast at this point :-|.

---
title: Intro to Docker and govuk-docker
section: Docker
layout: manual_layout
parent: "/manual.html"
last_reviewed_on: 2019-07-01
review_in: 6 months
owner_slack: "#govuk-developers"
---

## Intro
If you are new to Docker here's a quick intro in the context of the GOV.UK stack.

### Install Docker

*   Download Docker [https://www.docker.com/get-started](https://www.docker.com/get-started)
*   Docker for Mac intro [https://docs.docker.com/docker-for-mac/](https://docs.docker.com/docker-for-mac/)

### Video explaining the concepts in Docker

[https://www.youtube.com/watch?v=i7yoXqlg48M](https://www.youtube.com/watch?v=i7yoXqlg48M) - first 30 minutes

## Learning docker with content-publisher
We'll run a few things here to get the [content-publisher] up and running in Docker. This is a convoluted example but it will help to explain the concepts involved.

### Run a container from an image

> * Make sure you have docker running:

Reviewer: ```suggestion
Make sure you have docker running:
```

Author: Same thing about bullet points as a step

```shell
$mac docker --version
Docker version 18.09.2, build 6247962
```

> * run from the root of the [content-publisher] project:

Reviewer: ```suggestion
Run from the root of the [content-publisher][content-publisher] project:
```

Author: Same thing about bullet points as a step

```shell
$mac cd ~/govuk/content-publisher

> $mac docker run -it --rm ubuntu

Reviewer: I wonder if we should skip the ubuntu example and just go straight for using the ruby image. It may be less confusing and/or time consuming (because each image will need to be downloaded).

```

> ##### What are the flags?

Reviewer: I don't think this needs to be a heading?

```shell
$mac docker run --help
...
-i, --interactive                    Keep STDIN open even if not attached
...
-t, --tty                            Allocate a pseudo-TTY
...
    --rm                             Automatically remove the container when it exits
```

#### What are Docker images and a container?
The `ubuntu` is the "image" that we specify. Using object oriented programming as an analogy, an image is like a "class". When we "instantiate" an instance of a class, these are called "containers".

Images are downloaded from the Docker Registry.

Another image is ruby:

```shell
docker run -it --rm ruby:2.6.3 bash
```

Here we get a ruby image and specifying which version we want. The `bash` argument at the end is passed to the docker run command. It will execute this inside the container, which means we get a bash shell.

### Volumes

> * Mount a volume by running:

Reviewer: ```suggestion
Mount a volume by running:
```

Author: Same thing about bullet points as a step

Reviewer: @chao-xian I find it a bit more difficult to read when mixing bullet points with code and other comments that span full width (not indented), but that's just my personal preference 🤷‍♀
